Update EIP-8037: Add access-list based deduplication and success/failure cases

CPerezz opened the pull request:
Adding 2 important fixes/clarifications to EIP-8037:

1. Success/Failure Gas Accounting: Properly distinguishes between successful and failed contract deployments. Post-execution costs (ACCOUNT_CREATION_COST, HASH_GAS, CODE_DEPOSIT_GAS) are now only charged on success, aligning with Ethereum's principle that gas should reflect actual resource consumption.

2. Access-List CodeHash Deduplication: Replaces database lookup mechanism with an access-list based approach to solve consensus divergence between different sync modes (full-sync vs snap-sync). Approximately 27,869 orphaned bytecodes exist in full-sync nodes with no live account links, making DB lookups non-deterministic.

Changes include:
- Explicit success/failure gas formulas with detailed rationale
- AccessTuple extension with checkCodeHash field
- Comprehensive rationale explaining sync-mode divergence problem
- Developer-facing specification with Go struct and JSON example
- Updated gas cost examples for all scenarios
- LaTeX formatting for mathematical notation

CPerezz committed:
Update EIP-8037: Add access-list based deduplication and success/failure gas accounting

This commit implements two critical fixes to EIP-8037:

1. Success/Failure Gas Accounting: Properly distinguishes between successful
   and failed contract deployments. Post-execution costs (ACCOUNT_CREATION_COST,
   HASH_GAS, CODE_DEPOSIT_GAS) are now only charged on success, aligning with
   Ethereum's principle that gas should reflect actual resource consumption.

2. Access-List CodeHash Deduplication: Replaces database lookup mechanism with
   an access-list based approach to solve consensus divergence between different
   sync modes (full-sync vs snap-sync). Approximately 27,869 orphaned bytecodes
   exist in full-sync nodes with no live account links, making DB lookups
   non-deterministic.

Changes include:
- Explicit success/failure gas formulas with detailed rationale
- AccessTuple extension with checkCodeHash field
- Comprehensive rationale explaining sync-mode divergence problem
- Developer-facing specification with Go struct and JSON example
- Updated gas cost examples for all scenarios
- LaTeX formatting for mathematical notation

misilva73 commented on EIPS/eip-8037.md:
````diff
@@ -31,62 +31,133 @@ The relationship we are seeing in this example is not linear as expected. This i
+  + EXECUTION_GAS
+```
+
+#### Access-List CodeHash Deduplication
````
Wondering if we should add 2930 to the required EIPs field 🤔

CPerezz replied:
Hmmm. I thought that this is only done if the EIP required isn't merged. 
Happy to add it. 

Maybe @jochem-brouwer can bring some advice here.

misilva73 commented on EIPS/eip-8037.md:
```diff
@@ -103,6 +174,76 @@ This proposal is consistent with the multidimensional gas metering introduced in
+4. Explicit opt-in: Transactions must explicitly indicate which addresses to check, preventing unexpected behavior and giving users control over gas optimization.
+5. Forward compatibility: Older clients that don't implement the feature simply never grant the discount, maintaining compatibility while newer clients can optimize. Same will happen with Wallet providers and TX builders.
+
+#### Access-List Extension Specification
```
This section shouldn't be under "rationale". We should create a section named "Reference Implementation" . Check  [EIP-1](https://eips.ethereum.org/EIPS/eip-1#what-belongs-in-a-successful-eip)

CPerezz commented on EIPS/eip-8037.md:
```diff
@@ -31,62 +31,133 @@ The relationship we are seeing in this example is not linear as expected. This i
+     - If duplicate: do NOT charge `CODE_DEPOSIT_GAS(L)`; link `codeHash` to `H`
+     - If not duplicate: charge `CODE_DEPOSIT_GAS(L)` where `CODE_DEPOSIT_GAS(L) = 1,900 × L` and persist `R` under `H`, then link
+4. **Failure paths** (REVERT, OOG/invalid during initcode, OOG during code deposit, or `L > MAX_CODE_SIZE`):
+   - Do NOT charge `ACCOUNT_CREATION_COST`, `HASH_GAS(L)`, or `CODE_DEPOSIT_GAS(L)`
```
@gballet is this the case? 

Aren't all clients creating the account after or before? They do as they want?

@misilva73 my idea here was that if we fail, we won't create the account. Note that without selfdestruct, it should not be possible to create the account if initcode reverts.
Nevertheless, happy to change.

gballet replied:
ok, I've made a mistake here: this is only true of pre-homestead: post-homestead, the thing is reverted. so we don't have this problem today. That makes sense, you don't want to prevent a further deployment of code at this address.

But the question still remains: if you oog during code hash computation, do you want to create the contract if things still happen? to be consistent with the current behavior, you should not create anything. It's unfair, because you paid for it, but it's simpler.

gballet commented on EIPS/eip-8037.md:
```diff
@@ -31,62 +31,133 @@ The relationship we are seeing in this example is not linear as expected. This i
+
+#### Success vs Failure Gas Accounting
+
+When a contract creation transaction or opcode (`CREATE`/`CREATE2`) is executed, gas is charged differently based on whether the deployment succeeds or fails. Given runtime bytecode `R` (length `L`) returned by initcode and `H = keccak256(R)`:
```
runtime and `R` are not very good names imo. I would drop "runtime", as the rest of the sentence specifies that this is not the initcode. And then rename `R` to `B` so that it matches the name and isn't confusing to read.

gballet commented on EIPS/eip-8037.md:
```diff
@@ -31,62 +31,133 @@ The relationship we are seeing in this example is not linear as expected. This i
+
+When a contract creation transaction or opcode (`CREATE`/`CREATE2`) is executed, gas is charged differently based on whether the deployment succeeds or fails. Given runtime bytecode `R` (length `L`) returned by initcode and `H = keccak256(R)`:
+
+1. **On opcode entry:** Always charge `CREATE_FIXED_COST` (212,800 gas)
```
```suggestion
1. **When opcode execution starts:** Always charge `CREATE_FIXED_COST` (212,800 gas)
```

gballet commented on EIPS/eip-8037.md:
```diff
@@ -31,62 +31,133 @@ The relationship we are seeing in this example is not linear as expected. This i
+   - Charge `HASH_GAS(L)` where `HASH_GAS(L) = 6 × ceil(L / 32)` to compute `H`
+   - Perform deduplication check (see Access-List CodeHash section below):
+     - If duplicate: do NOT charge `CODE_DEPOSIT_GAS(L)`; link `codeHash` to `H`
+     - If not duplicate: charge `CODE_DEPOSIT_GAS(L)` where `CODE_DEPOSIT_GAS(L) = 1,900 × L` and persist `R` under `H`, then link
```
use `CODE_DEPOSIT_GAS` instead of 1900

CPerezz committed:
Address review comments for EIP-8037

Changes made:
1. Updated intro to explicitly mention bytecode deduplication discount
2. Added EIP-2930 to requires field (extends AccessTuple structure)
3. Reverted parameter names to match execution specs (GAS_CREATE, etc.)
4. Clarified forward compatibility: post-fork all clients enforce identical
   behavior; pre-fork behavior only differs between pre/post-fork, not
   between different client implementations
5. Moved Access-List Extension Specification to Reference Implementation
   section (per EIP-1 structure)
6. CRITICAL: Fixed W set building semantics with comprehensive clarification:
   - W is built from EXISTING contracts at transaction START
   - Only addresses that exist in state with deployed code contribute
   - Added example showing how T_A deploys, then T_B references T_A
   - Clarified edge case: same-block duplicate deployments both pay full
     cost if neither references existing contract (extremely rare, not
     worth special handling)
   - Updated both Specification and Reference Implementation sections

The most important fix addresses Maria's concern about W set confusion,
making it crystal clear that newly deployed contracts within a transaction
do NOT automatically add to W - only pre-existing state matters.

CPerezz commented on EIPS/eip-8037.md:
```diff
@@ -103,6 +174,76 @@ This proposal is consistent with the multidimensional gas metering introduced in
+2. No reverse index requirement: Unlike other approaches, this doesn't require maintaining a `codeHash → account` reverse index, which would add significant complexity and storage overhead.
+3. Leverages existing infrastructure: Builds on EIP-2930 access lists and EIP-2929 access costs, requiring minimal protocol changes.
+4. Explicit opt-in: Transactions must explicitly indicate which addresses to check, preventing unexpected behavior and giving users control over gas optimization.
+5. Forward compatibility: Older clients that don't implement the feature simply never grant the discount, maintaining compatibility while newer clients can optimize. Same will happen with Wallet providers and TX builders.
```
I removed the sentence as was super confusing. What I meant is that txs (wallets) that don't use the AccessList trick will still incurr into the prev gas model. 

So this was understood as pre-post fork. But I realize is just easier to remove it.

CPerezz commented on EIPS/eip-8037.md:
```diff
@@ -31,62 +31,133 @@ The relationship we are seeing in this example is not linear as expected. This i
+When a contract creation transaction or opcode (`CREATE`/`CREATE2`) is executed, gas is charged differently based on whether the deployment succeeds or fails. Given runtime bytecode `R` (length `L`) returned by initcode and `H = keccak256(R)`:
+
+1. **On opcode entry:** Always charge `CREATE_FIXED_COST` (212,800 gas)
+2. **During initcode execution:** Charge `EXECUTION_GAS` equal to gas consumed by the initcode
```
Applied the change from @gballet 

Also, agree the formula is more complex. But `exec_initcode_costs` implicitly accounts for mem expansion and call data costs. As this isn't something we are changing.

CPerezz commented:
> Optimisation on duplicated bytecode is something that we can do even now. That change should be contained in its own EIP and discussed separately from the gas cost increase.

Asked the rest of EL clients. Once I get answers from the rest will act accordingly. I thought about including it because this is part of the "gas harmonization cost" idea. But, indeed, this is something we can already do now..

gballet commented on EIPS/eip-8037.md:
```diff
@@ -103,6 +182,32 @@ This proposal is consistent with the multidimensional gas metering introduced in
+
+The access-list approach solves the bytecode duplication check problem by:
+
+1. Deterministic behavior: Only code hashes explicitly included in the transaction's access list are considered for deduplication. This makes the result independent of local database state.
```
but you would still need to load the account to make sure that the hash is the one you specify in the list

spencer-tb commented:
Can we close this or move it to a new EIP?

misilva73 commented:
We should close it, @spencer-tb. Can you do it, @CPerezz ?
